bentoml/configuration/containers.py

```diff
@@ -0,0 +1,170 @@
+# Copyright 2020 Atalaya Tech, Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import multiprocessing
+import os
+
+from deepmerge import always_merger
+from dependency_injector import containers, providers
+from schema import Schema, SchemaError, And, Or
+
+from bentoml.configuration import config
+from bentoml.exceptions import BentoMLConfigException
+from bentoml.utils.ruamel_yaml import YAML
+
+LOGGER = logging.getLogger(__name__)
+
+SCHEMA = Schema({
+    "api_server": {
+        "port": And(int, lambda port: port > 0),
+        "enable_microbatch": bool,
+        "run_with_ngrok": bool,
+        "enable_swagger": bool,
+        "enable_metrics": bool,
+        "enable_feedback": bool,
+        "max_request_size": And(int, lambda size: size > 0),
+        "workers": Or(And(int, lambda workers: workers > 0), None),
+        "timeout": And(int, lambda timeout: timeout > 0),
+    },
+    "marshal_server": {
+        "max_batch_size": And(int, lambda size: size > 0),
+        "max_latency": And(int, lambda latency: latency > 0),
+        "workers": Or(And(int, lambda workers: workers > 0), None),
+        "request_header_flag": str,
+    },
+    "yatai": {
+        "url": Or(str, None),
+    },
+    "tracing": {
+        "zipkin_api_url": Or(str, None),
+    },
+    "instrument": {
+        "namespace": str,
+    },
+})
```

Comment:
Compared the cerberus and schema libraries. I prefer the schema library because it is more pythonic and allow python logic in the definition of schema. However, cerberus has better documentation. Happy to discuss more.

Reply:
Both of them are good libraries, hard to go wrong.  I think `schema`'s more pythonic approach is better in our case.

For allowing python logic, we are using Cerberus this way: https://github.com/bentoml/BentoML/blob/93cd4b1b6ef8488473736399a979e5800cee2075/bentoml/yatai/validator/deployment_pb_validator.py#L154

Seems like `Schema` is cleaner to deal with.